tests/ws/server.go

```diff
@@ -79,16 +177,31 @@ func getWebsocketHandlerCDP(
 		done := make(chan struct{})
 		writeCh := make(chan cdproto.Message)
 
-		// Read loop
 		go func() {
+			read := func(conn *websocket.Conn) (*cdproto.Message, error) {
```

Comment:
Yeah, but we don't need to reuse them as they are part of the CDP handler, and we can always pull them out if we need to. I don't think that will be a big problem.

This is OK:

```golang
go func() {
	for {
		select {
		case <-done:
			return
		default:
		}

		_, buf, err := conn.ReadMessage()
		if err != nil {
			close(done)
			return
		}

		var msg cdproto.Message
		decoder := jlexer.Lexer{Data: buf}
		msg.UnmarshalEasyJSON(&decoder)
		if err := decoder.Error(); err != nil {
			close(done)
			return
		}

		if msg.Method != "" && cmdsReceived != nil {
			*cmdsReceived = append(*cmdsReceived, msg.Method)
		}

		fn(conn, &msg, writeCh, done)
	}
}()

go func() {
	for {
		select {
		case msg := <-writeCh:
			encoder := jwriter.Writer{}
			msg.MarshalEasyJSON(&encoder)
			if err := encoder.Error; err != nil {
				return
			}
			writer, err := conn.NextWriter(websocket.TextMessage)
			if err != nil {
				return
			}
			if _, err := encoder.DumpTo(writer); err != nil {
				return
			}
			if err := writer.Close(); err != nil {
				return
			}
		case <-done:
			return
		}
	}
}()
```

But it's not as understandable and easy-to-follow as the following at first glance. And as a side-bonus, we don't need to call `close(done)` twice:

```golang

go func() {
	read := func(conn *websocket.Conn) (*cdproto.Message, error) {
		_, buf, err := conn.ReadMessage()
		if err != nil {
			return nil, err
		}

		var msg cdproto.Message
		decoder := jlexer.Lexer{Data: buf}
		msg.UnmarshalEasyJSON(&decoder)
		if err := decoder.Error(); err != nil {
			return nil, err
		}

		return &msg, nil
	}

	for {
		select {
		case <-done:
			return
		default:
		}

		msg, err := read(conn)
		if err != nil {
			close(done)
			return
		}

		if msg.Method != "" && cmdsReceived != nil {
			*cmdsReceived = append(*cmdsReceived, msg.Method)
		}

		fn(conn, msg, writeCh, done)
	}
}()

go func() {
	write := func(conn *websocket.Conn, msg *cdproto.Message) {
		encoder := jwriter.Writer{}
		msg.MarshalEasyJSON(&encoder)
		if err := encoder.Error; err != nil {
			return
		}

		writer, err := conn.NextWriter(websocket.TextMessage)
		if err != nil {
			return
		}
		if _, err := encoder.DumpTo(writer); err != nil {
			return
		}
		if err := writer.Close(); err != nil {
			return
		}
	}

	for {
		select {
		case msg := <-writeCh:
			write(conn, &msg)
		case <-done:
			return
		}
	}
}()
```

I thought we agreed on not discussing trivial style issues as long as it works and is not terrible :)

Reply:
Sure, I was just wondering about the motivation to move it, since the standalone function was fine IMO.

> I thought we agreed on not discussing trivial style issues as long as it works and is not terrible :)

We did? OK, noted :smile: